enhancements/testing/ephemeral-agent-installer-e2e-tests.md

```diff
@@ -0,0 +1,337 @@
+isn't an easy way to stop the current frameworks from using IPI and Assisted Installer
+after the VM creation.
+
+
+#### Launch a VM with the Ephemeral Agent Based Installer Image
+
+As a developer, I need to launch a VM with the built ephemeral agent based installer
+image so that I can test and validate this image.
+
+Both existing frameworks don't have an easy way to do this. Which makes sense because
+they predate the ephemeral agent based installer project. There also isn't an easy way
+to use an existing routine to just launch a generic ISO image which would have been
+something we could have re-purposed. 
+
+### API Extensions
+
+Not applicable at this time.
+
+### Implementation Details/Notes/Constraints [optional]
+
+### Risks and Mitigations
+
+The direction of this proposal is now to extend and refactor parts of dev-scripts
+for the agent-based installer.
+
+** There will be now a third team using dev-scripts for slightly different purposes.
+Which means the risk of breaking things in the central framework now is larger and
+has a larger "splash zone."
+
+We can mitigate this for the new ephemeral agent installer by adding tests into the
+dev-scripts repository itself. The purpose of this would be to find breakages in the
+tests across the different teams.
+
+## Design Details
+
+### Overall Components of the e2e Framework
+
+Using the general flow of an e2e test we can sort the overall components of the e2e framework.
+A general flow of an e2e test has the following phases:
+
+* Pre Environment Setup: 
+** Setting up VMs, networks, locations of images, image registries, etc
+* Build: 
+** Building the agent-based installer ISO and container images associated with it
+* Test:
+**  Test scenarios with pass/failure outputs
+* Reporting:
+** Reports comprised of test results
+* Environment Teardown
+** Removing VMs, networks, etc to start fresh again
+
+#### Pre Environment Setup 
+
+* The e2e framework will be a collection of Ansible playbooks and roles
+* Using roles to define reusable steps scoped to units of work
+* Use playbooks to set up common e2e workflows and chains for the project
+
+In discussions regarding this enhancement we identified the dev-scripts has components that are
+a part of the Pre Environment Setup and Environment Teardown phases of the framework. We will
+move forward with adding and potentially refactoring scripts in this repository for these phases.
+
+Thinking about a general pattern of how to organize these things in the repository:
+
+```markdown
+playbooks/
+  example-e2e-workflow.yml
+roles/
+    common/
+        tasks/
+        handlers/
+        library/
+        files/
+        templates/
+        vars/
+        defaults/
+        meta/
+    example_role/
+        tasks/
+        defaults/
+        meta/
+```
+
+This is pretty open ended and meant to be. I just believe that focusing on making
+reusable roles here for the steps is key.
```

Comment:
Is this level of refactoring a prerequisite for making dev-scripts do something useful for the agent-based installer team? Or is this a longer term goal to clean up the current implementation?

Reply:
My plan is to add the new things for agent-installer using these patterns. In some cases that would mean a refactoring of things in the current implementation in the future. Iterative improvements over time.